### ISSUE TITLE
[RFC] Animation System

### ISSUE THREAD

**Issue by minecrawler** [amethyst/amethyst#214]
While the animation system might not be the highest priority right now, it still is important to put down some thoughts well ahead of time, imho.

# Motivation

A virtual world is nothing more than us trying to synthesize the real world. So we copy things we see. For instance, we create objects which live in the world; stuff like crates, tables and humans. However, it does not simply stop there! Certain kinds of objects need to change their shape in order to seem real. That's what we call "animations". In classic game engines, there are animation systems based on hand-made animations and some calculation on top of that, like blend trees and transitions. However, I believe that it is impossible to create all the variety needed to reflect the real world. As such, we have to start being more generic about more parts of the engine! Instead of loading animations and playing them back, why not generate the animations themselves and adapt based on the programming (instead of presets)?

# Implementation

The game engine should be able to create a variety of animations and blend them together programmatically. It should be possible to do gazing, talking, facial expression, locomotion, jumping, reaching/grabbing/touching, head movements (such as nodding), gesturing, eye saccades, blinking and breathing (and more), all of which should work together with the physics engine. A requirement for those is a defined skeleton structure. Additional parameters can be passed to achieve a variable animation, so male/female differences, as well as characteristic animations can be achieved.

Apart from the above mentioned innovative features, a way to configure bones and load animations should be available, too, which can be used by the above described system to feed the skeleton changes back.

The implementation could be achieved with a humanoid animation component a user just has to assign to a humanoid entity. There might be an "animation" component (required for the generic system) as well, which is used for animations in general.

# Drawbacks

Lots of work! I think, we might get some inspiration from [Smart Body](http://smartbody.ict.usc.edu), though.

# Unresolved Questions

Does this feature fit into the scope of the engine or should it be handled in the game instead of the engine? What about non-humanoid creatures, like dogs, horses, etc?

**Comment by OvermindDL1** [comment]
To me this really sounds per-game rather than engine unless you are just talking about a generic skeleton system and/or vertex blends?

**Comment by torkleyy** [comment]
I like the idea of generating animations, but why not use Smart Body? Or isn't this something the modeling software should do?

**Comment by minecrawler** [comment]
@OvermindDL1  Well, it should be quite generic for humanoid, that's why I noted that it should be parameterized in order to get different effects. While it might benefit my game if I kinda implement it, all other games can benefit from it if it is in the engine or at least a related crate; also people who know what they do could improve it to a degree I cannot even dream of :D

@torkleyy SmartBody is C++, unaware of the ECS, brings its own physics engine and is LGPL licensed.   So, for opensource games or commercial ones (which pay money for a different license), it might be possible to use it, but would be less than suboptimal. Modeling software does not have the capabilities to do so, because the animation is calculated in real time.

That's why I think it might be nice to develop that kind of system, make it opensource (MPL-2.0 or similar) and implement it in a way which fits Amethyst. Amethyst is a very innovative, data-driven game engine, which might be the right place to put innovative ideas like that. Also, Amethyst contains everything needed and the described animation system is generic in nature and can fit any simulation with humanoid actors. It takes away burden from 3D artists and allows for easy prototyping. Just rig a mesh with certain bones and register them with the system, and you have a functional actor!

If you are interested in such a system, but don't think it is a high priority, I could look into implementing it, since I really want to use that kind of system in my game anyway. I might take a while, though^^

**Comment by palodequeso** [comment]
I think the key takeaways here are a skeletal animation system, with support for animation blending, and inverse kinematics.

**Comment by agmcleod** [comment]
@palodequeso as a user of spine, i would love that :D

**Comment by msiglreith** [comment]
At the first glance, this looks quite targeted to a certain problem domain (character animation) but I assume the implementation requires several building blocks like @palodequeso mentioned: IK, general interpolation for animation, blending, constraint solving etc.
Interesting would to see how these parts can be coupled with other systems like transform, physics and our ECS imo.

**Comment by nchashch** [comment]
TL;DR: It would be nice to handle animation using a generic `Path<P: VectorSpace>` struct or trait, and it would be nice to make it possible to set an animation rig configuration using generalized coordinates (so instead of setting a whole transformation matrix of a car wheel you would set a single angle, defining it's orientation).

# Generic animation system:

I think it would be a good idea to have a lower level + more generic layer underneath the rigging system. Something like:
```rust
struct Path<P: VectorSpace> { ... }
```
Could define an animation curve, meaning you can call say `my_path.sample(t)` and get back an interpolated value of type `P`. `Path` would store a `Vec` of `(point, time)` pairs and an interpolation level (linear/quadratic/cubic/...). Basically it would be a spline approximating some function of type `time -> point` in vector space `P`. AFAIK it is a standard practice in animation to use splines for interpolation.

`P` having a `VectorSpace` bound means that:
1. You can add values of type `P` and the sum would still have type `P`.
2. You can multiply values of type `P` by a number and still get back a value of type `P`.
(AFAIK it is enough to construct bezier curves/splines)

`cgmath` has a `VectorSpace` trait, and it looks like it maps to the mathematical concept pretty well.

Reasons for `P` being generic:
1. You might want to animate different kinds of things that still satisfy vector space constraint. i.e. Color RGB vectors, position vectors, quaternions.
2. It might allow us to handle rigging in a generic way (more on this in next section).
3. It shouldn't be too hard to do animation blending in this way (summing two weighed `Path`s?).

# How to handle rigging in this way:

All possible configurations of an animation rig should form a vector space. For example if your rig is defined by transformation matricies for individual parts for your model and parent/child relations to tie it together, defining addition of rigs/multiplication by a scalar for such rigs should be straightforward, you just add corresponding matricies/multiply all matricies in the rig by a scalar. Then you can just define a `Path<AnimationRig>`.

# Making transform system more powerful:

On top of "matricies with parent/child relations" (which is already implemented) we could build a more high level transform system, based on generalized coordinates. Which would make handling transformations more tractable. Also it should make physics less hard in (constraints are handled automatically).

Generalized coordinates should still form a vector space, so say `Path<Configuration>` should work.

Example:
We want to set configuration of a [double pendulum](https://en.wikipedia.org/wiki/Double_pendulum). With "matricies with parent/child relations" system, there would be 2 matricies and hence 32 numbers to set (for 2 4x4 matricies) or we would have to set positino/scale/orientation, besides constraints aren't enforced. With generalized coordinates you would set 2 angles (one for parent pendulum and one for child pendulum).

**Comment by nchashch** [comment]
Abstract: This is a description of a specific way to do animation using generalized coordinates.

Suppose we have a car in the game represented as a parent-child tree of meshes. Each node apart from a mesh has a 4x4 transformation matrix, which represents a transformation in it's local coordinate system, current transform `System` works somewhat like this.

For generalized coordinates animation, we can define a `CarParameters` struct for this car object and a `CarParameters` -> `ParentChildTree` function. The `CarParameters` struct would contain a small number of semantically significant parameters describing this car. For example:
```rust
struct CarParameters {
    position: [f32; 3],
    orientation: [f32; 4],
    wheel_angle: [f32; 4],
    steering_wheel_angle: f32,
    door_angle: [f32; 4],
    wiper_angle: [f32; 2],
}
```
Note that this `CarParameters` form of car state description is easier to work with than a tree of 4x4 matrices with a node for every piece of a car. And also note that you can easily add `CarParameters` structs and multiply them by scalars (you just add corresponding fields or multiply all fields by a scalar). This is a good thing because it allows us to interpolate between `CarParameters` structs using splines. For example, an animator then would be able to define a `time` -> `CarParameters` function, by setting keyframes and choosing an interpolation method (linear, quadratic, cubic).

The `CarParameters` -> `ParentChildTree` function would transform a high-level `CarParameters` form of car description into a low level `ParentChildTree` form. It would be written by the person creating this particular game object (in this case a car), and it would encode all the physical constraints of a car. A wheel can only turn around a single axis. With `CarParameters` this constraint is handled by the `CarParameters` -> `ParentChildTree` function, and so it is hidden from the car API consumer. If you work with a 4x4 matrix of a wheel directly it is car API consumer's responsibility to handle this constraint.

At runtime the engine would sample the `time` -> `CarParameters` function, get the `ParentChildTree` via `CarParameters` -> `ParentChildTree` function, and pass the raw 4x4 matrices from `ParentChildTree` to the renderer.

**Comment by Rhuagh** [comment]
Here is my initial proposal for an animation system. This is a simple skeletal animation system, but swapping out `SkeletalAnimations`, `SkeletalAnimationControl` and `SkeletalAnimationSystem` for something else and you can do any type of animation you like basically.

```
/// Defines the joints entities for lookup purposes in rendering
pub struct Skeleton {
    joints: Vec<Entity>,
}

impl Storage for Skeleton {
    type Storage = DenseVecStorage<Skeleton>;
}

/// Single animation
pub struct SkeletalAnimation {
    sampler : AnimationSampler,
    channels: Vec<(Entity, AnimationChannel)>
}

/// All animations for an entity, should be attached to the root node of the mesh hierarchy
pub struct SkeletalAnimations {
    animations : Vec<SkeletalAnimation>,
}

impl Storage for SkeletalAnimations {
    type Storage = DenseVecStorage<SkeletalAnimations>;
}

pub struct SkeletalRestPose {
    position: [f32; 3],
    rotation: [f32; 4],
    scale: [f32; 3],
}

/// A single skeletal joint, should be on it's own entity with a `LocalTransform` and `Transform`
pub struct SkeletalJoint {
    inverse_bind_matrix: [[f32; 4]; 4],
    rest_pose: SkeletalRestPose, // used for resetting `LocalTransform` after animation ends
}

impl Storage for SkeletalJoint {
    type Storage = DenseVecStorage<SkeletalJoint>;
}

/// Attach to an entity with `Animations` to run the animation given
pub struct AnimationControl {
    active_index: usize, // index into `Animations::animations`
    start_time: u64,
}

impl Storage for AnimationControl {
    type Storage = DenseVecStorage<AnimationControl>;
}

/// Use `AnimationControl` to compute new `LocalTransform` for all joints
pub struct SkeletalAnimationSystem {}
```

**Comment by Rhuagh** [comment]
Updated proposal after feedback.

**Comment by Rhuagh** [comment]
```
#ifdef GL_ES
    precision highp float;
#endif

attribute vec3 a_position;

attribute vec4 a_joint;
attribute vec4 a_weight;

uniform mat4 u_jointMat[MAX_JOINTS];

uniform mat4 u_modelViewMatrix;
uniform mat4 u_projectionMatrix;

void main(void) 
{
    mat4 skinMat = 
        a_weight.x * u_jointMat[int(a_joint.x)] +
        a_weight.y * u_jointMat[int(a_joint.y)] +
        a_weight.z * u_jointMat[int(a_joint.z)] +
        a_weight.w * u_jointMat[int(a_joint.w)];
    vec4 pos = u_modelViewMatrix * skinMat * vec4(a_position,1.0);
    gl_Position = u_projectionMatrix * pos;
}
```

Adding this for future reference.

**Comment by jojolepro** [comment]
Anyone familiar with the animation systems can list what features we are missing?

**Comment by jojolepro** [comment]
bump.
Need the missing feature list to know if we try to get this done for 0.9 or delay it more.

**Comment by Rhuagh** [comment]
Animation controllers, blend trees, neural network spawned dynamic animations, usability

**Comment by erlend-sh** [comment]
Did anyone ever try to make Piston’s 
https://github.com/PistonDevelopers/skeletal_animation work with Amethyst just to try *some* kind of animation system in practice?

**Comment by fhaynes** [comment]
Closing this due to lack of activity and no assignee. We'll revisit Animation in 2019 planning. Captured here in the roadmap: https://openproject.amethyst-engine.org/projects/render/work_packages/62/activity
